sdk/spring/spring-security-azure-oauth2-client/src/main/java/com/azure/spring/security/oauth2/client/endpoint/AddScopeOAuth2AuthorizationCodeGrantRequestEntityConverter.java

```diff
@@ -0,0 +1,120 @@
+ *                 result.setRequestEntityConverter(
+ *                     new AddScopeOAuth2AuthorizationCodeGrantRequestEntityConverter("client1", scopes);
+ *             return result;
+ *         }
+ *     }
+ * </code>
+ * </pre>
+ * @author RujunChen
+ * @see OtherClientRefreshTokenOAuth2AuthorizedClientProvider
+ * @since 4.0
+ */
+public class AddScopeOAuth2AuthorizationCodeGrantRequestEntityConverter
+    extends OAuth2AuthorizationCodeGrantRequestEntityConverter {
+
+    private final String targetClientRegistrationId;
+    private final Set<String> scopes;
+
+    public AddScopeOAuth2AuthorizationCodeGrantRequestEntityConverter(
+        String targetClientRegistrationId, Set<String> scopes) {
+        this.targetClientRegistrationId = targetClientRegistrationId;
+        this.scopes = scopes;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public RequestEntity<?> convert(OAuth2AuthorizationCodeGrantRequest request) {
+        RequestEntity<?> requestEntity = super.convert(request);
+        Assert.notNull(requestEntity, "requestEntity can not be null");
+
+        HttpHeaders httpHeaders = getHttpHeaders();
+        Optional.of(requestEntity)
+                .map(HttpEntity::getHeaders)
+                .ifPresent(headers -> headers.forEach(httpHeaders::put));
+        MultiValueMap<String, String> body = (MultiValueMap<String, String>) requestEntity.getBody();
+        Assert.notNull(body, "body can not be null");
+        Optional.ofNullable(getHttpBody(request)).ifPresent(body::putAll);
+        return new RequestEntity<>(body, httpHeaders, requestEntity.getMethod(), requestEntity.getUrl());
+    }
+
+    /**
+     * Additional headers information used to record current client information.
+     * @return HttpHeaders
+     */
+    public HttpHeaders getHttpHeaders() {
+        HttpHeaders httpHeaders = new HttpHeaders();
+        httpHeaders.put("x-client-SKU", Collections.singletonList(ApplicationId.AZURE_SPRING_AAD));
+        httpHeaders.put("x-client-VER", Collections.singletonList(ApplicationId.VERSION));
+        httpHeaders.put("client-request-id", Collections.singletonList(UUID.randomUUID().toString()));
+        return httpHeaders;
+    }
+
+    /**
+     * Add "scope" parameter if the clientRegistration is targetClientRegistration
+     *
+     * @param request OAuth2AuthorizationCodeGrantRequest
+     * @return MultiValueMap
+     */
+    public MultiValueMap<String, String> getHttpBody(OAuth2AuthorizationCodeGrantRequest request) {
+        MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
+        String scopeValue = String.join(" ", isTargetClientRegistration(request)
+            ? scopes
+            : request.getClientRegistration().getScopes());
+        body.add("scope", scopeValue);
+        return body;
+    }
+
+    private boolean isTargetClientRegistration(OAuth2AuthorizationCodeGrantRequest request) {
+        return Optional.of(request)
```

Comment:
you really should STOP using this Optional api

Reply:
Why?
We already discussed this, `Optional` can avoid NullPointerException. And Some other open-source project already use `Optional` api. It's a good practice to use `Optional` api.

Even in this repo, you can find a lot of usages except in `sdk/spring` folder.

Reply:
Optional is fine API to **consider** if it is part of your API, but in this case it isn't part of your API. You could write the same code without any Optional and still avoid NPE, and even the code below is not safe from NPE. I don't see the need for it in this case.

Reply:
> You could write the same code without any Optional and still avoid NPE

Yes, here are 2 options:

### Option 1: 
```
    private boolean isTargetClientRegistration(OAuth2AuthorizationCodeGrantRequest request) {
        return request.getClientRegistration().getRegistrationId().equals(targetClientRegistrationId);
    }
```
If you are pretty sure that `request` is not null, and `request.getClientRegistration()` is not null, and `request.getClientRegistration().getRegistrationId()` is not null. If we are not sure that all these are not null, we need Option 2:

### Option 2
```
    private boolean isTargetClientRegistration(OAuth2AuthorizationCodeGrantRequest request) {
        return request != null
            && request.getClientRegistration() != null
            && request.getClientRegistration().getRegistrationId() != null
            && request.getClientRegistration().getRegistrationId().equals(targetClientRegistrationId);
    }
```

Compare to option 2, I prefer current code:

### Current code
```
    private boolean isTargetClientRegistration(OAuth2AuthorizationCodeGrantRequest request) {
        return Optional.of(request)
                       .map(AbstractOAuth2AuthorizationGrantRequest::getClientRegistration)
                       .map(ClientRegistration::getRegistrationId)
                       .map(id -> id.equals(targetClientRegistrationId))
                       .orElse(false);
    }
```
So we do not need to consider which step will cause NPE, we can concentrate on our own logic.

Reply:
please, for our code, STOP using this optional, I'd more prefer the first one
Optional API is not designed for that